File: csharp/robot-name/RobotName.cs
using System;
using System.Collections.Generic;

public class Robot
{
    private string name = "";
    private const string LETTERS = "QWERTZUIOPASDFGHJKLYXCVBNM";
    private static readonly HashSet<string> NAMES = [];
    public string Name => name;

    private static readonly Random rnd = new();

    public Robot() => Reset();

    public void Reset()
    {
        string newName;
        do
        {
            newName = "";
            newName += LETTERS[rnd.Next(0, LETTERS.Length)];
            newName += LETTERS[rnd.Next(0, LETTERS.Length)];
            newName += rnd.Next(0, 10);
            newName += rnd.Next(0, 10);
            newName += rnd.Next(0, 10);
        } while (NAMES.Contains(newName));
        NAMES.Add(newName);
        name = newName;
    }
}

File: kotlin/robot-name/src/main/kotlin/RobotName.kt
class Robot {

    companion object {
        private val usedNames = mutableSetOf<String>()
        private val letters = ('A'..'Z').toList()
        private val digits = ('0'..'9').toList()
    }

    var name: String = generateUniqueName()
        private set

    fun reset() {
        name = generateUniqueName()
    }

    private fun generateUniqueName(): String {
        while (true) {
            val newName = buildString {
                repeat(2) { append(letters.random()) }
                repeat(3) { append(digits.random()) }
            }

            if (usedNames.add(newName)) {
                return newName
            }
        }
    }
}
